dns_server/src/crawler/tests/mod.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod mock;
+
+use std::{net::SocketAddr, time::Duration};
+
+use p2p::testing_utils::P2pTestTimeGetter;
+use tokio::sync::mpsc;
+
+use crate::{
+    crawler::{
+        storage::{DnsServerStorageRead, DnsServerTransactional},
+        Crawler, CrawlerConfig,
+    },
+    dns_server::ServerCommands,
+};
+
+use self::mock::{
+    MockConnectivityHandle, MockNetworkingService, MockStateRef, MockSyncingMessagingHandle,
+};
+
+use super::storage_impl::DnsServerStorageImpl;
+
+fn test_crawler(
+    add_node: Vec<SocketAddr>,
+) -> (
+    Crawler<MockNetworkingService, DnsServerStorageImpl<storage::inmemory::InMemory>>,
+    MockStateRef,
+    mpsc::UnboundedReceiver<ServerCommands>,
+    P2pTestTimeGetter,
+) {
+    let (conn_tx, conn_rx) = mpsc::unbounded_channel();
+    let add_node = add_node.iter().map(ToString::to_string).collect();
+    let crawler_config = CrawlerConfig {
+        add_node,
+        network: [1, 2, 3, 4],
+        p2p_port: 3031,
+    };
+
+    let state = MockStateRef {
+        crawler_config: crawler_config.clone(),
+        online: Default::default(),
+        connected: Default::default(),
+        connection_attempts: Default::default(),
+        conn_tx,
+    };
+
+    let conn = MockConnectivityHandle {
+        state: state.clone(),
+        conn_rx,
+    };
+    let sync = MockSyncingMessagingHandle {};
+
+    let storage = storage::inmemory::InMemory::new();
+    let store = DnsServerStorageImpl::new(storage).unwrap();
+
+    let (command_tx, command_rx) = mpsc::unbounded_channel();
+
+    let crawler =
+        Crawler::<MockNetworkingService, _>::new(crawler_config, conn, sync, store, command_tx)
+            .unwrap();
+
+    let time_getter = P2pTestTimeGetter::new();
+
+    (crawler, state, command_rx, time_getter)
+}
+
+async fn advance_time(
```

Comment:
Can you please provide description for this function that explains how it works?

Reply:
Sure, done.